src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -71,14 +71,132 @@ const rpc::PlacementGroupTableData &GcsPlacementGroup::GetPlacementGroupTableDat
 std::string GcsPlacementGroup::DebugString() const {
   std::stringstream stream;
   stream << "placement group id = " << GetPlacementGroupID() << ", name = " << GetName()
-         << ", strategy = " << GetStrategy();
+         << ", strategy = " << GetStrategy()
+         << ", creator job dead: " << placement_group_table_data_.creator_job_dead()
+         << ", creator actor dead: " << placement_group_table_data_.creator_actor_dead();
   return stream.str();
 }
 
 rpc::Bundle *GcsPlacementGroup::GetMutableBundle(int bundle_index) {
   return placement_group_table_data_.mutable_bundles(bundle_index);
 }
 
+const ActorID GcsPlacementGroup::GetCreatorActorId() const {
+  return ActorID::FromBinary(placement_group_table_data_.creator_actor_id());
+}
+
+const JobID GcsPlacementGroup::GetCreatorJobId() const {
+  return JobID::FromBinary(placement_group_table_data_.creator_job_id());
+}
+
+void GcsPlacementGroup::MarkCreatorJobDead() {
+  placement_group_table_data_.set_creator_job_dead(true);
+}
+
+void GcsPlacementGroup::MarkCreatorActorDead() {
+  placement_group_table_data_.set_creator_actor_dead(true);
+}
+
+bool GcsPlacementGroup::IsPlacementGroupRemovable() const {
+  return placement_group_table_data_.creator_job_dead() &&
+         placement_group_table_data_.creator_actor_dead();
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+bool RegisteredPlacementGroupIndex::Emplace(
+    const PlacementGroupID &placement_group_id,
+    std::shared_ptr<GcsPlacementGroup> placement_group) {
+  if (placement_groups_.find(placement_group_id) != placement_groups_.end()) {
+    return false;
+  }
+  RAY_CHECK(placement_groups_.emplace(placement_group_id, placement_group).second);
+
+  // Fill in actor index.
+  const ActorID &actor_id = placement_group->GetCreatorActorId();
+  if (!actor_id.IsNil()) {
+    if (placement_groups_by_actor_id_.find(actor_id) ==
+        placement_groups_by_actor_id_.end()) {
+      placement_groups_by_actor_id_[actor_id] = PlacementGroupMap();
+    }
+    auto it = placement_groups_by_actor_id_.find(actor_id);
+    it->second.emplace(placement_group_id, placement_group);
```

Comment:
auto it = placement_groups_by_actor_id_.find(actor_id);
if (it == placement_groups_by_actor_id_.end()) {
  it = placement_groups_by_actor_id_.emplace(placement_group_id, placement_group).first;
}

Reply:
This class will be removed.